src/Build/Graph/GraphBuilder.cs

```diff
@@ -110,28 +111,74 @@ private static IReadOnlyCollection<ProjectGraphNode> GetGraphRoots(IReadOnlyColl
+
             foreach (var parsedProject in allParsedProjects)
             {
+                var currentNode = parsedProject.Value.GraphNode;
+
+                var requiresTransitiveProjectReferences = _projectInterpretation.RequiresTransitiveProjectReferences(currentNode.ProjectInstance);
+
                 foreach (var referenceInfo in parsedProject.Value.ReferenceInfos)
                 {
-                    ErrorUtilities.VerifyThrow(
-                        allParsedProjects.ContainsKey(referenceInfo.ReferenceConfiguration),
-                        "all references should have been parsed");
-
-                    parsedProject.Value.GraphNode.AddProjectReference(
+                    // Always add direct references.
+                    currentNode.AddProjectReference(
                         allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode,
                         referenceInfo.ProjectReferenceItem,
                         edges);
+
+                    // Add transitive references only if the project requires it.
+                    if (requiresTransitiveProjectReferences)
+                    {
+                        foreach (var transitiveProjectReference in GetTransitiveProjectReferences(allParsedProjects[referenceInfo.ReferenceConfiguration]))
+                        {
+                            currentNode.AddProjectReference(
+                                transitiveProjectReference,
+                                new ProjectItemInstance(
+                                    project: currentNode.ProjectInstance,
+                                    itemType: ProjectInterpretation.TransitiveReferenceItemName,
+                                    includeEscaped: referenceInfo.ReferenceConfiguration.ProjectFullPath,
+                                    directMetadata: null,
+                                    definingFileEscaped: currentNode.ProjectInstance.FullPath
+                                ),
+                                edges);
+                        }
+                    }
+                }
+            }
+
+            IEnumerable<ProjectGraphNode> GetTransitiveProjectReferences(ParsedProject parsedProject)
+            {
+                if (transitiveReferenceCache.TryGetValue(parsedProject.GraphNode, out var cachedTransitiveReferences))
+                {
+                    return cachedTransitiveReferences;
+                }
+                else
+                {
+                    var transitiveReferences = new List<ProjectGraphNode>();
+
+                    foreach (var referenceInfo in parsedProject.ReferenceInfos)
+                    {
+                        transitiveReferences.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode);
+                        transitiveReferences.AddRange(GetTransitiveProjectReferences(allParsedProjects[referenceInfo.ReferenceConfiguration]));
```

Comment:
Worth worrying about stack depth here? I hope not . . .

Reply:
GraphBuilder.DetectCycles.VisitNode traverses the entire graph and is already recursive, so if that one's recursive, then this new one won't increase the risk of stack overflows.
If we ever find a graph that busts the stack, the temporary workaround for that repo would be to temporarily increase the stack size with `editbin.exe` while we change all recursive traversals to iterative.

If you really want to go iterative now, I'd do it as a separate PR and not block this one on it.

Reply:
Nah, "not worse than before" is ok.